### PR TITLE
ARROW-3136: [C++] Clean up public API

### PR DESCRIPTION
**Pull request by pitrou**
Move some exported functions and classes into `arrow::util` and `arrow::internal` namespaces.
Also add and improve some docstrings for public API members.

### CODE REVIEW

File: cpp/src/arrow/type_fwd.h [hunk 1]
```diff
@@ -135,26 +135,42 @@ using IntervalArray = NumericArray<IntervalType>;
...
+/// \brief Return a StringType instance
 std::shared_ptr<DataType> ARROW_EXPORT utf8();
+/// \brief Return a BinaryType instance
 std::shared_ptr<DataType> ARROW_EXPORT binary();
-
+/// \brief Return a Date32Type instance
 std::shared_ptr<DataType> ARROW_EXPORT date32();
+/// \brief Return a Date64Type instance
 std::shared_ptr<DataType> ARROW_EXPORT date64();
-std::shared_ptr<DataType> ARROW_EXPORT decimal_type(int precision, int scale);
```
Comment (pitrou) [inline cpp/src/arrow/type_fwd.h]: Note this function isn't defined anywhere AFAICT.

Comment (wesm) [reply cpp/src/arrow/type_fwd.h]: There is the `arrow::decimal(p, s)` so that's enough

**Review by wesm — APPROVED** [review]
+1, this all looks fine to me. Thanks for slogging through this, @pitrou! I will merge this soon to help prevent rebase conflicts

**Comment by pitrou** [comment]
The R task fails, I think it's unrelated but I'm not sure.

**Comment by wesm** [comment]
See https://github.com/apache/arrow/pull/2635. I will merge that once the build runs, then rebase this so we can check that R is OK after your changes

**Comment by pitrou** [comment]
AppVeyor build at https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/build/1.0.8109

**Comment by pitrou** [comment]
AppVeyor builds: https://ci.appveyor.com/project/pitrou/arrow/build/1.0.766 and https://ci.appveyor.com/project/pitrou/arrow/build/1.0.767. I'm gonna merge.

**Pull request closed without merging**